// rumble-charts/cjs/helpers/transforms/rotate.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../external/tslib/tslib.es6.js');

function rotate(props) {
    var series = props.series, seriesNormalized = props.seriesNormalized, minX = props.minX, maxX = props.maxX, maxY = props.maxY, minY = props.minY;
    var scaleX = props.scaleX, scaleY = props.scaleY;
    var paddingLeft = scaleX.paddingLeft, paddingRight = scaleX.paddingRight;
    var paddingTop = scaleY.paddingTop, paddingBottom = scaleY.paddingBottom;
    scaleX = tslib_es6.__assign({}, scaleX);
    scaleY = tslib_es6.__assign({}, scaleY);
    scaleX.paddingLeft = paddingTop;
    scaleX.paddingRight = paddingBottom;
    scaleX.swap = !scaleX.swap;
    scaleY.paddingTop = paddingLeft;
    scaleY.paddingBottom = paddingRight;
    scaleY.swap = !scaleY.swap;
    scaleY.direction = -1;
    return {
        series: series,
        seriesNormalized: seriesNormalized,
        maxX: maxX,
        maxY: maxY,
        minX: minX,
        minY: minY,
        scaleX: scaleX,
        scaleY: scaleY
    };
}

exports.rotate = rotate;
