// rumble-charts/cjs/helpers/transforms/sort.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../external/tslib/tslib.es6.js');
var isString = require('../isString.js');
var sortBy = require('../sortBy.js');

function sort(props, _a) {
    var _b = _a === void 0 ? {} : _a, direction = _b.direction;
    direction = (isString.isString(direction) ? direction.trim().toLowerCase() : 'asc');
    var seriesNormalized = props.seriesNormalized, maxX = props.maxX, maxY = props.maxY, minX = props.minX, minY = props.minY;
    var series = props.series;
    series = series === null || series === void 0 ? void 0 : series.map(function (series) {
        var _a;
        var newSeries = { data: [] };
        newSeries.data = sortBy.sortBy(series.data, 'y');
        if (direction === 'desc') {
            newSeries.data.reverse();
        }
        newSeries.data = (_a = newSeries.data) === null || _a === void 0 ? void 0 : _a.map(function (point, pointIndex) {
            var newPoint = {
                realX: point.x,
                x: pointIndex
            };
            return tslib_es6.__assign(tslib_es6.__assign({}, point), newPoint);
        });
        newSeries.data = sortBy.sortBy(newSeries.data, 'realX');
        return tslib_es6.__assign(tslib_es6.__assign({}, series), newSeries);
    });
    return {
        series: series,
        seriesNormalized: seriesNormalized,
        maxX: maxX,
        maxY: maxY,
        minX: minX,
        minY: minY
    };
}

exports.sort = sort;
