// rumble-charts/cjs/helpers/transforms/transpose.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../external/tslib/tslib.es6.js');

function transpose(props) {
    var series = props.series, seriesNormalized = props.seriesNormalized, minX = props.minX, maxY = props.maxY, minY = props.minY;
    var maxX = 0;
    var newSeries = [];
    series === null || series === void 0 ? void 0 : series.forEach(function (series, seriesIndex) {
        var _a;
        (_a = series.data) === null || _a === void 0 ? void 0 : _a.forEach(function (point, pointIndex) {
            newSeries[pointIndex] = newSeries[pointIndex] || { data: [] };
            maxX = Math.max(maxX, seriesIndex);
            newSeries[pointIndex].data[seriesIndex] = tslib_es6.__assign(tslib_es6.__assign({}, point), { realX: point.x, x: seriesIndex });
        });
    });
    return {
        series: newSeries,
        seriesNormalized: seriesNormalized,
        maxX: maxX,
        maxY: maxY,
        minX: minX,
        minY: minY
    };
}

exports.transpose = transpose;
