// rumble-charts/cjs/helpers/transforms/unstack.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../external/tslib/tslib.es6.js');
var omit = require('../omit.js');

function unstack(props) {
    var series = props.series, seriesNormalized = props.seriesNormalized, maxX = props.maxX, maxY = props.maxY, minX = props.minX, minY = props.minY;
    var newSeries = series
        ? series.map(function (series) {
            var _a;
            var newSeries = {
                data: (_a = series.data) === null || _a === void 0 ? void 0 : _a.map(function (point) {
                    var newPoint = omit.omit(point, ['y0', 'y']);
                    newPoint.y = point.y - point.y0;
                    return newPoint;
                })
            };
            return tslib_es6.__assign(tslib_es6.__assign({}, series), newSeries);
        }, [])
        : [];
    return {
        series: newSeries,
        seriesNormalized: seriesNormalized,
        maxX: maxX,
        maxY: maxY,
        minX: minX,
        minY: minY
    };
}

exports.unstack = unstack;
