// rumble-charts/cjs/helpers/colorFunc.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var d3Scale = require('d3-scale');
var d3ScaleChromatic = require('d3-scale-chromatic');
var isFunction = require('./isFunction.js');
var isString = require('./isString.js');

var schemeMap = {
    category10: d3ScaleChromatic.schemeCategory10,
    category20: [
        '#1f77b4', '#aec7e8', '#ff7f0e', '#ffbb78', '#2ca02c',
        '#98df8a', '#d62728', '#ff9896', '#9467bd', '#c5b0d5',
        '#8c564b', '#c49c94', '#e377c2', '#f7b6d2', '#7f7f7f',
        '#c7c7c7', '#bcbd22', '#dbdb8d', '#17becf', '#9edae5'
    ],
    category20b: [
        '#393b79', '#5254a3', '#6b6ecf', '#9c9ede', '#637939',
        '#8ca252', '#b5cf6b', '#cedb9c', '#8c6d31', '#bd9e39',
        '#e7ba52', '#e7cb94', '#843c39', '#ad494a', '#d6616b',
        '#e7969c', '#7b4173', '#a55194', '#ce6dbd', '#de9ed6'
    ],
    category20c: [
        '#3182bd', '#6baed6', '#9ecae1', '#c6dbef', '#e6550d',
        '#fd8d3c', '#fdae6b', '#fdd0a2', '#31a354', '#74c476',
        '#a1d99b', '#c7e9c0', '#756bb1', '#9e9ac8', '#bcbddc',
        '#dadaeb', '#636363', '#969696', '#bdbdbd', '#d9d9d9'
    ],
    accent: d3ScaleChromatic.schemeAccent,
    dark2: d3ScaleChromatic.schemeDark2,
    paired: d3ScaleChromatic.schemePaired,
    pastel1: d3ScaleChromatic.schemePastel1,
    pastel2: d3ScaleChromatic.schemePastel2,
    set1: d3ScaleChromatic.schemeSet1,
    set2: d3ScaleChromatic.schemeSet2,
    set3: d3ScaleChromatic.schemeSet3,
    tableau10: d3ScaleChromatic.schemeTableau10
};
var defaultScheme = schemeMap.category20;
var defaultSchemeName = 'category20';
function colorFunc(colors) {
    if (isFunction.isFunction(colors)) {
        return colors;
    }
    else if (!colors) {
        return d3Scale.scaleOrdinal(defaultScheme);
    }
    else if (isString.isString(colors)) {
        return d3Scale.scaleOrdinal(schemeMap[colors]);
    }
    else {
        return d3Scale.scaleOrdinal(colors);
    }
}

exports.colorFunc = colorFunc;
exports.defaultScheme = defaultScheme;
exports.defaultSchemeName = defaultSchemeName;
exports.schemeMap = schemeMap;
